qa/L0_backend_bls/bls.cc

```diff
@@ -0,0 +1,349 @@
+  TRITONSERVER_MemoryType output_memory_type_backend;
+  int64_t output_memory_id_backend;
+  const void* output_base_backend;
+  void* userp_backend;
+  for (size_t icount = 0; icount < 2; icount++) {
+    // Retrieve the corresponding TRITONSERVER_InferenceResponse object from
+    // 'futures'. The InferResponseComplete function sets the std::promise
+    // so that this thread will block until the response is returned.
+    completed_responses[icount] = futures[icount].get();
+    try {
+      THROW_IF_TRITON_ERROR(
+          TRITONSERVER_InferenceResponseError(completed_responses[icount]));
+    }
+    catch (const BLSBackendException& bls_exception) {
+      LOG_MESSAGE(TRITONSERVER_LOG_ERROR, bls_exception.what());
+
+      if (completed_responses[icount] != nullptr) {
+        LOG_IF_ERROR(
+            TRITONSERVER_InferenceResponseDelete(completed_responses[icount]),
+            "Failed to delete inference response.");
+      }
+      return;
+    }
+    // Retrieve outputs from 'completed_responses'.
+    // Extract OUTPUT0 from the 'addsub_python' and OUTPUT1 from the
+    // 'addsub_tf' model to form the final inference response object.
+    // Get all the information about the output tensor.
+    RESPOND_AND_SET_NULL_IF_ERROR(
+        response, TRITONSERVER_InferenceResponseOutput(
+                      completed_responses[icount], icount, &output_name_server,
+                      &output_datatype_server, &output_shape_server,
+                      &dims_count_server, &output_base_server,
+                      &output_byte_size_server, &output_memory_type_server,
+                      &output_memory_id_server, &userp_server));
+
+    // Create an output tensor in the final response with
+    // the information retrieved above.
+    TRITONBACKEND_Output* output;
+    RESPOND_AND_SET_NULL_IF_ERROR(
+        response,
+        TRITONBACKEND_ResponseOutput(
+            *response, &output, output_name_server, output_datatype_server,
+            output_shape_server, dims_count_server));
+
+    // Get a buffer that holds the tensor data for the output.
+    // We request a buffer in CPU memory but we have to handle any returned
+    // type. If we get back a buffer in GPU memory we just fail the request.
+    void* output_buffer;
+    output_memory_type_server = TRITONSERVER_MEMORY_CPU;
+    RESPOND_AND_SET_NULL_IF_ERROR(
+        response, TRITONBACKEND_OutputBuffer(
+                      output, &output_buffer, output_byte_size_server,
+                      &output_memory_type_server, &output_memory_id_server));
+    if (output_memory_type_server == TRITONSERVER_MEMORY_GPU) {
+      RESPOND_AND_SET_NULL_IF_ERROR(
+          response, TRITONSERVER_ErrorNew(
+                        TRITONSERVER_ERROR_INTERNAL,
+                        "failed to create output buffer in CPU memory"));
+    }
+    // Validate new backen APIs, ensuring that the retrieved values match those
+    // retrieved by the server
```

Comment:
Instead of adding a majority copy of this `bls.cc` file into the server, can we just augment the backend example and mention it's just for validation/demonstration?

Something along the lines of this:

```cpp
// NOTE: This function is just to demonstrate TRITONBACKEND equivalents of TRITONSERVER output
// getters, and verify their correctness. It is not necessary for this backend example.
TRITONSERVER_Error*
ValidateBackendGetters() {
  /* your existing additions, but RETURN_IF_ERROR instead of RESPOND_IF_ERROR everywhere */
}


void
BLSExecutor::ConstructFinalResponse() {
  ...
  
  // Optional: demonstrate backend output getter correctness, not necessary for this example.
  RESPOND_AND_SET_NULL_IF_ERROR(ValidateBackendGetters(...));
  
  ...
}
```
This way you're just adding the 1 new helper function to the example, keeping the code in one place in the example, and not duplicating it in server.

Also, we can add a comment that the helpers can be used for optimizations (original intent of the feature I believe?)

Reply:
@fpetrini15 @Tabrizian what do you think?

Reply:
@Tabrizian since this is an example, it might be helpful to add the comment as rmccorm4 suggests, but it might be better to demonstrate the optimization. How feasible/is this the appropriate model to demonstrate the possibility of the response optimization?

Reply:
I figured out an easier way to unit test the backend APIs without having to create a separate backend. Please see the PR here: https://github.com/triton-inference-server/core/pull/249/files#diff-50be12abfe003477dda43701f231fad704847749ad914a52868dbf8afc508fb9

Reply:
Added unit testing for the backend APIs: https://github.com/triton-inference-server/core/pull/247